README.md
# Site Editor Classic Features

Welcome to the Site Editor Classic Features repository on GitHub. Here you can browse the source, discuss/open issues and keep track of the development.

If you are not a developer, we recommend to [download Site Editor Classic Features](https://wordpress.org/plugins/fse-classic/) from WordPress directory.

**Todo** Change the link if we change the name of the plugin before publishing on WordPress directory.

## [What this plugin is for](#fse-classic)

Currently this plugin adds the legacy widgets block in any kind of block editor, more specifically the Site Editor (FSE) introduced in WordPress 5.9.

By reintroducing the legacy widgets block, you can use some of the legacy widgets as `Navigation Menu` for example.

However with blocks theme, like Twenty Twenty-Two, you can't update menus anymore because this kind of theme doesn't support menus.

This is why this plugin reintroduces the WordPress admin menus to be able to manage menus you want to use with the legacy widgets block.

This also allows to use external plugins legacy widgets which might not have already been converted to blocks.

## [Pre-requisites](#pre-requisites)

Before starting, make sure that you have the following software installed and working on your computer:

1. A local [WordPress](https://wordpress.org/support/article/how-to-install-wordpress/) (5.9 or later) instance
2. [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) to clone the FSE Classic repository (or your fork of the FSE Classic repository).
3. [Node.js](https://nodejs.org/en/download/) which provides [NPM](https://docs.npmjs.com/). They are both required by [Webpack](https://webpack.js.org/guides/getting-started/) that FSE Classic uses to build and minify CSS and javascript files. We recommend to install Node.js LTS version.
4. [Composer](https://getcomposer.org/doc/00-intro.md) because FSE Classic uses its autoloader to work and it is required to install development tools such as PHP CodeSniffer that ensures your code follows coding standards.

## [How to set up Site Editor Classic Features](#how-to-setup-fse-classic)

The simplest way is to clone locally this repository and build it directly in your local WordPress instance by following the steps below:

1. Go to your local WordPress instance wp-content/plugins/ folder:<br/>
`cd your/local/wordpress/path/wp-content/plugins`
2. Clone there the FSE Classic repository (or your fork) from GitHub:<br/>
`git clone https://github.com/polylang/fse-classic.git`
3. Go to your local fse-classic clone folder from there: `cd fse-classic`
4. Run the composer command: `composer build`
5. Activate FSE Classic as if you had installed it from WordPress.org:<br/>

**Note**: we recommend Windows users to use `Git Bash` provided with [Git for Windows](https://git-scm.com/download/win) instead of the command or powershell terminals.


changelog.txt
= 1.1 (2023-10-10) =

* Require at least PHP 7.0.
* Move from Travis to GitHub Actions.
* Tested up to WordPress 6.4.

= 1.0 (2022-12-21) =
* Initial release


src/Plugin.php
<?php
/**
 * Main plugin class.
 * php version 5.6
 *
 * @package WP_Syntex\FSE_Classic
 */

namespace WP_Syntex\FSE_Classic;

defined( 'ABSPATH' ) || exit;

/**
 * Main plugin class.
 *
 * @since 1.0
 */
final class Plugin {

	/**
	 * Suffix for minified assets or not. Depending on the debug environment.
	 *
	 * @var string
	 */
	private $assets_suffix;

	/**
	 * Constructor.
	 *
	 * @since 1.0
	 *
	 * @return void
	 */
	public function __construct() {
		$this->assets_suffix = defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ? '' : '.min';
	}

	/**
	 * Plugin init.
	 *
	 * @since 1.0
	 *
	 * @return void
	 */
	public function init() {
		/**
		 * Fires before the plugin init.
		 *
		 * @since 1.0
		 *
		 * @param Plugin $plugin Main class instance.
		 */
		do_action_ref_array( 'fse_classic_before_init', [ &$this ] );

		if ( wp_is_block_theme() ) {
			// Loads the plugin only if the theme is a block theme.
			$this->addHooks();
		}

		/**
		 * Fires after the plugin init.
		 *
		 * @since 1.0
		 *
		 * @param Plugin $plugin Main class instance.
		 */
		do_action_ref_array( 'fse_classic_init', [ &$this ] );
	}

	/**
	 * Adds hooks.
	 *
	 * @since 1.0
	 *
	 * @return void
	 */
	private function addHooks() {
		add_action( 'admin_print_styles', [ $this, 'adminStyles' ] );
		add_action( 'wp_enqueue_scripts', [ $this, 'classicMenuStyles' ] );
		add_action( 'admin_print_scripts', [ $this, 'adminScripts' ] );
		add_action( 'admin_print_footer_scripts', [ $this, 'adminFooterScripts' ] );
		add_action( 'admin_footer', [ $this, 'adminFooterWidgets' ] );
		add_action( 'enqueue_block_editor_assets', [ $this, 'editorAssets' ] );
		add_action( 'after_setup_theme', [ $this, 'supportMenusAndWidgets' ] );
	}

	/**
	 * Loads the widget scripts and registers the Legacy Widget block via JS.
	 *
	 * @since 1.0
	 *
	 * @return void
	 */
	public function editorAssets() {
		wp_enqueue_script( 'wp-widgets' );
		wp_add_inline_script( 'wp-widgets', 'wp.widgets.registerLegacyWidgetBlock()' );
		wp_enqueue_style( 'wp-widgets' );
	}

	/**
	 * Loads the classic menu style.
	 *
	 * @since 1.0
	 *
	 * @return void
	 */
	public function classicMenuStyles() {
		wp_enqueue_style(
			'fse-classic-menu',
			plugins_url( '/public/css/menu' . $this->assets_suffix . '.css', FSE_CLASSIC_FILE ),
			[],
			FSE_CLASSIC_VERSION
		);
	}

	/**
	 * Runs the widget styles action in the block editor.
	 *
	 * @since 1.0
	 *
	 * @return void
	 */
	public function adminStyles() {
		if ( $this->isBlockEditor() ) {
			do_action( 'admin_print_styles-widgets.php' ); // phpcs:ignore WordPress.NamingConventions.ValidHookName.UseUnderscores
		}
	}

	/**
	 * Runs various widget actions in the block editor.
	 *
	 * @since 1.0
	 *
	 * @return void
	 */
	public function adminScripts() {
		if ( $this->isBlockEditor() ) {
			do_action( 'load-widgets.php' ); // phpcs:ignore WordPress.NamingConventions.ValidHookName.UseUnderscores
			do_action( 'widgets.php' ); // phpcs:ignore WordPress.NamingConventions.ValidHookName.UseUnderscores
			do_action( 'sidebar_admin_setup' );
			do_action( 'admin_print_scripts-widgets.php' ); // phpcs:ignore WordPress.NamingConventions.ValidHookName.UseUnderscores
		}
	}

	/**
	 * Runs the footer widget scripts action in the block editor.
	 *
	 * @since 1.0
	 *
	 * @return void
	 */
	public function adminFooterScripts() {
		if ( $this->isBlockEditor() ) {
			do_action( 'admin_print_footer_scripts-widgets.php' ); // phpcs:ignore WordPress.NamingConventions.ValidHookName.UseUnderscores
		}
	}

	/**
	 * Runs the widgets footer action in the block editor.
	 *
	 * @since 1.0
	 *
	 * @return void
	 */
	public function adminFooterWidgets() {
		if ( $this->isBlockEditor() ) {
			do_action( 'admin_footer-widgets.php' ); // phpcs:ignore WordPress.NamingConventions.ValidHookName.UseUnderscores
		}
	}

	/**
	 * Allows to display the menus admin page.
	 *
	 * @since 1.0
	 *
	 * @return void
	 */
	public function supportMenusAndWidgets() {
		add_theme_support( 'menus' );
	}

	/**
	 * Check if the current screen runs the block editor.
	 *
	 * @since 1.0
	 *
	 * @return bool True if the current screen runs the block editor, false otherwise.
	 */
	private function isBlockEditor(): bool {
		$current_screen = get_current_screen();

		if ( ! empty( $current_screen ) ) {
			return $current_screen->is_block_editor();
		}

		return false;
	}
}


webpack.config.js
/**
 * @package Polylang
 */

 /**
 * External dependencies
 */

const path = require( 'path' );
const glob = require( 'glob' ).sync;
const { transformJsEntry, transformCssEntry } = require( '@wpsyntex/polylang-build-scripts' );

function configureWebpack( options ){
	const mode = options.mode;
	const isProduction = mode === 'production' || false;
	console.log('Webpack mode:', mode);
	console.log('isProduction:', isProduction);
	console.log('dirname:', __dirname);

	const commonFoldersToIgnore = [
		'node_modules/**',
		'vendor/**',
		'tmp/**',
		'webpack/**',
		'**/public/**',
	];

	const jsFileNamesToIgnore = [
		'js/lib/**',
		'**/*.config.js',
		'**/*.min.js',
	];

	const jsFileNames = glob( '**/*.js', { 'ignore': [ ...commonFoldersToIgnore, ...jsFileNamesToIgnore ] } ).map( filename => `./${ filename }`);
	console.log( 'js files to minify:', jsFileNames );

	const jsFileNamesEntries = [
		...jsFileNames.map( transformJsEntry( path.resolve( __dirname ) + '/public/js', true ) ),
		...jsFileNames.map( transformJsEntry( path.resolve( __dirname ) + '/public/js', false ) )
	]

	const cssFileNames = glob( '**/*.css', { 'ignore': [ ...commonFoldersToIgnore, '**/*.min.css' ] } ).map( filename => `./${ filename }`);
	console.log( 'css files to minify:', cssFileNames );

	// Prepare webpack configuration to minify css files to source folder as target folder and suffix file name with .min.js extension.
	const cssFileNamesEntries = cssFileNames.map( transformCssEntry( path.resolve( __dirname ) + '/public/css', isProduction ) );

	// Make webpack configuration.
	const config = [
		...jsFileNamesEntries, // Add config for js files.
		...cssFileNamesEntries, // Add config for css files.
	];

	return config;
}

module.exports = ( env, options ) => {
	return configureWebpack( options );
}



legacy-features-fse.php
<?php
/**
 * Legacy features for FSE
 * php version        5.6
 *
 * @package           WP_Syntex\Legacy_Features_for_FSE
 * @author            WP SYNTEX
 * @license           GPL-3.0-or-later
 *
 * @wordpress-plugin
 * Plugin Name:       Legacy Features for FSE
 * Plugin URI:        https://polylang.pro
 * Description:       Legacy Features for FSE.
 * Version:           1.0
 * Requires at least: 5.9
 * Requires PHP:      5.6
 * Author:            WP SYNTEX
 * Author uri:        https://polylang.pro
 * License:           GPL v3 or later
 * License URI:       https://www.gnu.org/licenses/gpl-3.0.txt
 *
 * Copyright 2022 WP SYNTEX
 *
 * This program incorporates work from the plugin Legacy Widget Block
 * Copyright 2022 Justin Tadlock (email : justintadlock@gmail.com)
 * Legacy Widget Block is released under the GPL V2 or later.
 */

namespace WP_Syntex\Legacy_Features_for_FSE;

defined( 'ABSPATH' ) || exit;

define( 'LFFF_VERSION', '1.0' );

define( 'LFFF_FILE', __FILE__ );

if ( file_exists( __DIR__ . '/vendor/autoload.php' ) ) {
	require_once __DIR__ . '/vendor/autoload.php';
}

( new Plugin() )->init();


fse-classic.php
<?php
/**
 * Site Editor Classic Features
 * php version        5.6
 *
 * @package           WP_Syntex\FSE Classic
 * @author            WP SYNTEX
 * @license           GPL-3.0-or-later
 *
 * @wordpress-plugin
 * Plugin Name:       Site Editor Classic Features
 * Plugin URI:        https://polylang.pro
 * Description:       Allows to use classic widgets and menus in Site Editor.
 * Version:           1.1
 * Requires at least: 5.9
 * Requires PHP:      7.0
 * Author:            WP SYNTEX
 * Author uri:        https://polylang.pro
 * Text Domain:       fse-classic
 * License:           GPL v3 or later
 * License URI:       https://www.gnu.org/licenses/gpl-3.0.txt
 *
 * Copyright 2022 WP SYNTEX
 *
 * This program incorporates work from the plugin Legacy Widget Block
 * Copyright 2022 Justin Tadlock (email : justintadlock@gmail.com)
 * Legacy Widget Block is released under the GPL V2 or later.
 */

namespace WP_Syntex\FSE_Classic;

defined( 'ABSPATH' ) || exit;

define( 'FSE_CLASSIC_VERSION', '1.1' );

define( 'FSE_CLASSIC_FILE', __FILE__ );

if ( file_exists( __DIR__ . '/vendor/autoload.php' ) ) {
	require_once __DIR__ . '/vendor/autoload.php';
}

( new Plugin() )->init();
